pkg/cableengine/syncgatewaystatus.go

```diff
@@ -0,0 +1,161 @@
+
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/log"
+)
+
+const GatewayUpdateIntervalSeconds = 5
+const GatewayStaleTimeoutSeconds = GatewayUpdateIntervalSeconds * 3
+const updateTimestampAnnotation = "update-timestamp"
+
+func (i *engine) syncGatewayStatus() {
+
+	i.Lock()
+	defer i.Unlock()
+
+	klog.V(log.TRACE).Info("Running syncGatewayStatus()")
+
+	gatewayObj, err := i.generateGatewayObject()
+	if err != nil {
+		klog.Errorf("generating gateway object from driver connections: %s", err)
+		return
+	}
+
+	existingGw, err := i.getLastSyncedGateway(gatewayObj.Name)
+
+	// log and stop for any error different to a not-found error
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("trying to read existing gateway from k8s api: %s", err)
+		return
+	}
+
+	gwClient := i.clientset.SubmarinerV1().Gateways(i.namespace)
+
+	if err != nil && errors.IsNotFound(err) {
+		klog.V(log.TRACE).Infof("Gateway object needs creation: %+v", gatewayObj)
+		_, err = gwClient.Create(gatewayObj)
+		if err != nil {
+			klog.Errorf("creating Gateway object: %s", err)
+			return
+		}
+	} else {
+		if !reflect.DeepEqual(gatewayObj.Status, existingGw.Status) {
+			klog.V(log.TRACE).Infof("Gateway object needs an update: %+v", gatewayObj)
+			existingGw.Status = gatewayObj.Status
+			existingGw.Labels = gatewayObj.Labels
+			existingGw.Annotations = gatewayObj.Annotations
+
+			gw, err := gwClient.Update(existingGw)
+			if err != nil {
+				klog.Errorf("updating Gateway object: %s", err)
+				return
+			} else {
+				klog.V(log.TRACE).Infof("Gateway updated correctly: %+v", gw)
+			}
+		} else {
+			klog.V(log.TRACE).Info("Gateway object didn't need an update")
+		}
+	}
+
+	if gatewayObj.Status.HAStatus == v1.HAStatusActive {
+		err := i.cleanupStaleGatewayEntries()
+		if err != nil {
+			klog.Errorf("cleaning up stale gateway entries: %s", err)
+		}
+	}
+}
+
+func (i *engine) cleanupStaleGatewayEntries() error {
+	gwClient := i.clientset.SubmarinerV1().Gateways(i.namespace)
+	gateways, err := gwClient.List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, gw := range gateways.Items {
+		stale, err := isGatewayStale(gw)
+		if err != nil {
+			// In this case we don't want to stop the cleanup loop and just log it
+			klog.Errorf("error processing stale gateway: %+v , %s", gw, err)
+		}
+		if stale {
+			err := gwClient.Delete(gw.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				// In this case we don't want to stop the cleanup loop and just log it
+				klog.Errorf("error deleting stale gateway: %+v, %s", gw, err)
+			}
+			klog.Warningf("deleted stale gateway: %s, didn't report for %d seconds",
+				gw.Name, GatewayStaleTimeoutSeconds)
+		}
+	}
+	return nil
+}
+
+func isGatewayStale(gateway v1.Gateway) (bool, error) {
+
+	timestamp, ok := gateway.ObjectMeta.Annotations[updateTimestampAnnotation]
+	if !ok {
+		return true, fmt.Errorf("update-timestamp annotation not found")
```

Comment:
nit: can you use updateTimestampAnnotation instead of hardcoding the string "update-timestamp" (including below)

Reply:
makes sense, I'll update it.